File: backend/app/api/routers/reports.py
from datetime import date

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.schemas.reports import CashflowReport, CategoriesReport, TaxEstimateReport
from app.application.services.report_service import ReportService
from app.core.db import get_db
from app.core.dependencies import get_current_user_id

router = APIRouter(prefix="/reports", tags=["reports"])


@router.get("/cashflow", response_model=CashflowReport)
async def cashflow_report(
    date_from: date | None = None,
    date_to: date | None = None,
    session: AsyncSession = Depends(get_db),
    user_id: str = Depends(get_current_user_id),
):
    # Не задаём даты по умолчанию — сервис возьмёт диапазон по фактическим транзакциям
    service = ReportService(session)
    return await service.cashflow(user_id, date_from, date_to)


@router.get("/categories", response_model=CategoriesReport)
async def categories_report(
    date_from: date | None = None,
    date_to: date | None = None,
    session: AsyncSession = Depends(get_db),
    user_id: str = Depends(get_current_user_id),
):
    # Не задаём даты по умолчанию — сервис возьмёт диапазон по фактическим транзакциям
    service = ReportService(session)
    return await service.by_categories(user_id, date_from, date_to)


@router.get("/tax-estimate", response_model=TaxEstimateReport)
async def tax_estimate_report(
    date_from: date | None = None,
    date_to: date | None = None,
    session: AsyncSession = Depends(get_db),
    user_id: str = Depends(get_current_user_id),
):
    # Не задаём даты по умолчанию — сервис возьмёт диапазон по фактическим транзакциям
    service = ReportService(session)
    return await service.tax_estimate(user_id, date_from, date_to)

File: frontend/src/shared/api/reports.ts
import { api } from "./client"

export interface CashflowReport {
  date_from: string
  date_to: string
  total_inflow: number
  total_outflow: number
  net: number
  by_date: { date: string; net: number }[]
}

export interface CategoriesReport {
  date_from: string
  date_to: string
  by_category: Record<string, { inflow: number; outflow: number }>
}

export interface TaxEstimateReport {
  disclaimer: string
  date_from: string
  date_to: string
  income: number
  rate: number
  estimated_tax: number
}

export const reportsApi = {
  cashflow: (params?: { date_from?: string; date_to?: string }) => {
    const search = new URLSearchParams()
    if (params?.date_from) search.set("date_from", params.date_from)
    if (params?.date_to) search.set("date_to", params.date_to)
    const q = search.toString()
    return api.get<CashflowReport>(`/reports/cashflow${q ? `?${q}` : ""}`)
  },
  categories: (params?: { date_from?: string; date_to?: string }) => {
    const search = new URLSearchParams()
    if (params?.date_from) search.set("date_from", params.date_from)
    if (params?.date_to) search.set("date_to", params.date_to)
    const q = search.toString()
    return api.get<CategoriesReport>(`/reports/categories${q ? `?${q}` : ""}`)
  },
  taxEstimate: (params?: { date_from?: string; date_to?: string }) => {
    const search = new URLSearchParams()
    if (params?.date_from) search.set("date_from", params.date_from)
    if (params?.date_to) search.set("date_to", params.date_to)
    const q = search.toString()
    return api.get<TaxEstimateReport>(`/reports/tax-estimate${q ? `?${q}` : ""}`)
  },
}
